broker/src/test/java/io/zeebe/broker/test/AtomixLogStorageRule.java

```diff
@@ -0,0 +1,203 @@
+      final TemporaryFolder temporaryFolder,
+      final int partitionId,
+      final UnaryOperator<RaftStorage.Builder> builder) {
+    this.temporaryFolder = temporaryFolder;
+    this.partitionId = partitionId;
+    this.builder = builder;
+  }
+
+  @Override
+  public void before() throws Throwable {
+    open();
+  }
+
+  @Override
+  public void after() {
+    close();
+  }
+
+  @Override
+  public CompletableFuture<Indexed<ZeebeEntry>> appendEntry(
+      final long lowestPosition, final long highestPosition, final ByteBuffer data) {
+    final Indexed<ZeebeEntry> entry =
+        raftLog
+            .writer()
+            .append(
+                new ZeebeEntry(
+                    0, System.currentTimeMillis(), lowestPosition, highestPosition, data));
+    raftLog.writer().commit(entry.index());
+
+    if (positionListener != null) {
+      positionListener.accept(findGreatestPosition(entry));
+    }
+
+    return CompletableFuture.completedFuture(entry);
+  }
+
+  @Override
+  public AtomixLogStorage get() {
+    return storage;
+  }
+
+  @Override
+  public Optional<ZeebeLogAppender> getAppender() {
+    return Optional.of(this);
+  }
+
+  @Override
+  public CompletableFuture<Void> compact(final long index) {
+    raftLog.compact(index);
+    return CompletableFuture.completedFuture(null);
+  }
+
+  @Override
+  public RaftLogReader create(final long index, final Mode mode) {
+    return raftLog.openReader(index, mode);
+  }
+
+  public void setPositionListener(final LongConsumer positionListener) {
+    this.positionListener = positionListener;
+  }
+
+  public void open() {
+    open(builder);
+  }
+
+  public void open(final UnaryOperator<RaftStorage.Builder> builder) {
+    final File directory;
+    close();
+
+    try {
+      directory = temporaryFolder.newFolder(String.format("atomix-%d", partitionId));
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    raftStorage = builder.apply(buildDefaultStorage()).withDirectory(directory).build();
+    raftLog = raftStorage.openLog();
+    snapshotStore = raftStorage.openSnapshotStore();
+    metaStore = raftStorage.openMetaStore();
+
+    storage = spy(new AtomixLogStorage(this, this, this));
+  }
+
+  public void close() {
+    Optional.ofNullable(raftLog).ifPresent(RaftLog::close);
+    Optional.ofNullable(snapshotStore).ifPresent(SnapshotStore::close);
+  }
+
+  public int getPartitionId() {
+    return partitionId;
+  }
+
+  public AtomixLogStorage getStorage() {
+    return storage;
+  }
+
+  public RaftStorage getRaftStorage() {
+    return raftStorage;
+  }
+
+  public RaftLog getRaftLog() {
+    return raftLog;
+  }
+
+  public SnapshotStore getSnapshotStore() {
+    return snapshotStore;
+  }
+
+  public MetaStore getMetaStore() {
+    return metaStore;
+  }
+
+  private RaftStorage.Builder buildDefaultStorage() {
+    return RaftStorage.builder()
+        .withFlushOnCommit()
+        .withStorageLevel(StorageLevel.MAPPED)
```

Comment:
I like the idea, but now we are testing different paths then we use in the normal run.